### PR TITLE
send frame in packet on demand

### PR DESCRIPTION
**Pull request by ZwX1616**
e.g. `ZMQ=1 SEND_REAR=1 XMIN=388 XMAX=776 YMIN=200 YMAX=500 SCALE=4 ./camerad`

### CODE REVIEW

File: selfdrive/camerad/cameras/camera_common.cc [hunk 1]
```diff
@@ -220,6 +220,30 @@ void fill_frame_data(cereal::FrameData::Builder &framed, const FrameMetadata &fr
   framed.setGainFrac(frame_data.gain_frac);
 }
 
+void fill_frame_image(cereal::FrameData::Builder &framed, uint8_t *dat, int w, int h, int stride) {
+  if (dat != nullptr) {
+    int scale = 1;
+    int x_min = 0; int y_min = 0; int x_max = w-1; int y_max = h-1;
+    if (getenv("XMIN")) x_min = atoi(getenv("XMIN"));
```
Comment (adeebshihadeh) [inline selfdrive/camerad/cameras/camera_common.cc:227]: check this only once

File: selfdrive/camerad/cameras/camera_common.cc [hunk 2]
```diff
@@ -24,6 +24,13 @@
 #include "common/util.h"
 #include "imgproc/utils.h"
 
+const int env_xmin = getenv("XMIN") ? atoi(getenv("XMIN")) : 0;
+const int env_xmax = getenv("XMAX") ? atoi(getenv("XMAX")) : -1;
+const int env_ymin = getenv("YMIN") ? atoi(getenv("YMIN")) : 0;
+const int env_ymax = getenv("YMAX") ? atoi(getenv("YMAX")) : -1;
+const int env_scale = getenv("SCALE") ? atoi(getenv("SCALE")) : 1;
+const char *env_send_front = getenv("SEND_FRONT");
```
Comment (adeebshihadeh) [inline selfdrive/camerad/cameras/camera_common.cc:32]:
```suggestion
const bool env_send_front = getenv("SEND_FRONT") != NULL;
```

File: selfdrive/camerad/cameras/camera_qcom2.cc [hunk 3]
```diff
@@ -34,6 +34,8 @@
 
 
 extern volatile sig_atomic_t do_exit;
+const char *env_send_rear = getenv("SEND_REAR");
+const char *env_send_wide = getenv("SEND_WIDE");
```
Comment (adeebshihadeh) [inline selfdrive/camerad/cameras/camera_qcom2.cc:38]: I'd put all these together in camera common.

**Comment by pd0wm** [comment]
Is this what we want? Won't loggerd completely choke on this?

Why not write a visionipc to zmq bridge?

**Comment by ZwX1616** [comment]
> Is this what we want? Won't loggerd completely choke on this?
> 
> Why not write a visionipc to zmq bridge?

From what I understand loggerd won't care if we use the ZMQ flag. The only downside is that the frame packets are diverted to somewhere else and thus don't get logged. But we will be debugging the camera(s) anyways, and can handle the msgs with the tool of choice (e.g. replay ui or fl ui).

**Pull request merged**